atomic_reactor/plugins/pre_inject_yum_repos.py

```diff
@@ -0,0 +1,265 @@
+"""
+Copyright (c) 2015-2022 Red Hat, Inc
+All rights reserved.
+
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+
+
+Pre build plugin which injects custom yum repositories in dockerfile.
+"""
+import os
+import shutil
+from collections import defaultdict
+from io import StringIO
+from urllib.parse import urlparse
+
+from atomic_reactor.config import get_koji_session
+from atomic_reactor.constants import YUM_REPOS_DIR, RELATIVE_REPOS_PATH, INSPECT_CONFIG, \
+    PLUGIN_RESOLVE_COMPOSES_KEY
+from atomic_reactor.dirs import BuildDir
+from atomic_reactor.plugin import PreBuildPlugin
+from atomic_reactor.util import is_scratch_build, map_to_user_params, \
+    allow_repo_dir_in_dockerignore, render_yum_repo, get_platforms
+from atomic_reactor.utils.yum import YumRepo
+
+
+class InjectYumReposPlugin(PreBuildPlugin):
+    key = "inject_yum_repos"
+    is_allowed_to_fail = False
+
+    args_from_user_params = map_to_user_params(
+        "target:koji_target",
+        "repourls:yum_repourls",
+    )
+
+    def __init__(self, workflow, target=None, inject_proxy=None):
+        """
+        constructor
+
+        :param workflow: DockerBuildWorkflow instance
+        :param target: string, koji target to use as a source
+        :param inject_proxy: set proxy server for this repo
+        """
+        super().__init__(workflow)
+        self.target = target
+
+        self.repourls = {}
+        self.inject_proxy = inject_proxy
+        self.yum_repos = defaultdict(list)
+        self.allowed_domains = self.workflow.conf.yum_repo_allowed_domains
+        self.include_koji_repo = False
+        self._builder_ca_bundle = None
+        self._ca_bundle_pem = None
+        self._dockerfile = None
+        self.platforms = get_platforms(workflow)
+
+    def validate_yum_repo_files_url(self):
+        if not self.allowed_domains:
+            return
+        errors = []
+
+        for platform in self.platforms:
+            for repourl in self.repourls.get(platform, []):
+                repo_domain = urlparse(repourl).netloc
+                if repo_domain not in self.allowed_domains:
+                    errors.append('Yum repo URL {} is not in list of allowed domains: {}'
+                                  .format(repourl, self.allowed_domains))
```

Comment:
We expect some of the urls to be duplicated between platforms, don't we? Maybe it would make sense to de-duplicate them before checking them?

Reply:
de-deduplicated